server/src/main/java/org/opensearch/index/translog/transfer/TranslogCheckpointSnapshot.java

```diff
@@ -0,0 +1,136 @@
+        long generation,
+        long minTranslogGeneration,
+        Path translogPath,
+        Path checkpointPath,
+        Long translogChecksum,
+        Long checkpointChecksum,
+        Checkpoint checkpoint,
+        long checkpointGeneration
+    ) {
+        this.primaryTerm = primaryTerm;
+        this.generation = generation;
+        this.minTranslogGeneration = minTranslogGeneration;
+        this.translogPath = translogPath;
+        this.checkpointPath = checkpointPath;
+        this.translogChecksum = translogChecksum;
+        this.checkpointChecksum = checkpointChecksum;
+        this.checkpoint = checkpoint;
+        this.checkpointGeneration = checkpointGeneration;
+    }
+
+    public String getTranslogFileName() {
+        return translogPath.getFileName().toString();
+    }
+
+    public String getCheckpointFileName() {
+        return checkpointPath.getFileName().toString();
+    }
+
+    public long getTranslogFileContentLength() throws IOException {
+        try (FileChannel fileChannel = FileChannel.open(translogPath, StandardOpenOption.READ)) {
+            return fileChannel.size();
+        }
+    }
+
+    public long getCheckpointFileContentLength() throws IOException {
+        try (FileChannel fileChannel = FileChannel.open(checkpointPath, StandardOpenOption.READ)) {
+            return fileChannel.size();
+        }
+    }
+
+    public long getGeneration() {
+        return generation;
+    }
+
+    public long getPrimaryTerm() {
+        return primaryTerm;
+    }
+
+    public long getCheckpointGeneration() {
+        return checkpointGeneration;
+    }
+
+    TransferFileSnapshot getTranslogFileSnapshot() throws IOException {
+        return new TranslogFileSnapshot(primaryTerm, generation, translogPath, translogChecksum);
+    }
+
+    TransferFileSnapshot getCheckpointFileSnapshot() throws IOException {
+        return new CheckpointFileSnapshot(primaryTerm, generation, minTranslogGeneration, checkpointPath, checkpointChecksum);
+    }
+
+    TransferFileSnapshot getTranslogFileSnapshotWithMetadata() throws IOException {
+        Map<String, String> metadata = createCheckpointDataAsObjectMetadata();
+        TransferFileSnapshot translogFileSnapshot = getTranslogFileSnapshot();
+        translogFileSnapshot.setMetadata(metadata);
+        return translogFileSnapshot;
+    }
+
+    private Map<String, String> createCheckpointDataAsObjectMetadata() throws IOException {
+
+        byte[] fileBytes = Checkpoint.createCheckpointBytes(checkpointPath, checkpoint);
+
+        // Do checksum validation here.
+        TranslogCheckedContainer translogCheckedContainer = new TranslogCheckedContainer(fileBytes);
+        Long calculatedChecksum = translogCheckedContainer.getChecksum();
+
+        if (checkpointChecksum != null && !checkpointChecksum.equals(calculatedChecksum)) {
+            throw new TranslogUploadFailedException("Checksum validation of checkpoint file failed for translog file:");
+        }
```

Comment:
As I can see you are using an inmemory object to create the byte array. We dont need to rerun the checksum again here.